
<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.4" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs-1.1/textmate.css"
      type="text/css" />

<script src="../site_libs/highlightjs-1.1/highlight.js"></script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>M6A</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">M6A</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../highlights.html">Highlights</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
<li>
  <a href="../writeup.html">Writeup</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/gaow/m6a"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="ASE-calling-pipeline">ASE calling pipeline<a class="anchor-link" href="#ASE-calling-pipeline">&#182;</a></h1><p>Mostly based on Yanyu's work in 2017. The pipeline aligns <code>fastq</code> sequences to <code>bam</code> using <code>STAR</code>, then adjust the mapping via <code>WASP</code> to account for allele specificity, and finally call genotype and ASE via <code>QuASAR</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Overview-of-procedure">Overview of procedure<a class="anchor-link" href="#Overview-of-procedure">&#182;</a></h2><ol>
<li>Gather require resources</li>
<li>Align reads to genome</li>
<li>Remove biased reads</li>
<li>Genotype and ASE calling</li>
</ol>
<p>The pipeline is implemented in SoS as displayed in the rest of this SoS notebook. The pipeline can be executed from this notebook directly on local or remote computer (with caveats see below). For more information see <a href="https://github.com/vatlab/SOS">SoS website</a>.</p>
<ul>
<li>Some software used (<code>STAR</code>, <code>samtools</code>, <code>bedtools</code>) are loaded from online <code>docker</code> images. This is me being lazy to avoid installing these software on a local computer (which should not take much time anyways). But some cluster system, such as UChicago RCC, does not support <code>docker</code>. Also here I did not configure cluster job queues with <code>task</code> option. So this notebook is intended to run on a local workstation as is, though involves minimal changes to get it work distributively on cluster, given required software are installed there.</li>
</ul>
<p>To run the pipeline:</p>

<pre><code>sos run nb.ipynb hg19_reference
sos run nb.ipynb obtain_samples
sos run nb.ipynb default</code></pre>
<h3 id="Difference-from-Yanyu's-version">Difference from Yanyu's version<a class="anchor-link" href="#Difference-from-Yanyu's-version">&#182;</a></h3><p>Other than using a formal pipeline tool, differences are:</p>
<ol>
<li>Here samples are paired end reads -- this led to minor differences in <code>STAR</code> and <code>WASP</code> command parameters.</li>
<li>No techinical replicates in this problem -- skipped Yanyu's step of merging technical replicates to one bam file.</li>
</ol>
<h3 id="Lessons-learned:-a-retrospective-view">Lessons learned: a retrospective view<a class="anchor-link" href="#Lessons-learned:-a-retrospective-view">&#182;</a></h3><p>A list of cautions learned from running the pipeline:</p>
<ol>
<li>The <code>STAR</code> aligner is fast (compared to <code>tophat</code>) but requires over 30GB of memory. A desktop with 32GB of memory will not work for the <code>STAR</code> steps. I actually had to use a 64GB RAM desktop.</li>
<li><code>STAR</code>, when it fails (memory or not), will fail silently.</li>
<li>Preparation of <code>STAR</code> database takes long time, but it is portable as long as one changes the absolute path in <code>genomeParameters.txt</code> to relative (ie, remove all the directories from file names only keep the basename). This is what I did. So the resource prepared by <code>[star]</code> workflow (below) can be copied to use in other machines.</li>
</ol>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Settings-and-default-workflow">Settings and default workflow<a class="anchor-link" href="#Settings-and-default-workflow">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="n">cwd</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;~/Documents/m6A/Data/ASE&#39;</span><span class="p">)</span>
<span class="kn">parameter:</span> <span class="n">ncpu</span> <span class="o">=</span> <span class="mi">20</span>
<span class="c1"># Resource files</span>
<span class="n">resource_dir</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/hg19&quot;</span>
<span class="n">ref_fa</span> <span class="o">=</span> <span class="s2">&quot;hg19.fa&quot;</span>
<span class="n">ref_gtf</span> <span class="o">=</span> <span class="s1">&#39;Homo_sapiens.GRCh38.91.gtf.gz&#39;</span>
<span class="n">wasp_dir</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/WASP-master/mapping&quot;</span>
<span class="n">samtools_snp_list</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/020717_filtered_maf5_blacklist.final.bed&quot;</span>
<span class="c1"># Sample files</span>
<span class="n">sample_dir</span> <span class="o">=</span> <span class="n">f</span><span class="s2">&quot;{cwd:a}/samples&quot;</span>
<span class="kn">from</span> <span class="nn">collections</span> <span class="kn">import</span> <span class="n">OrderedDict</span>
<span class="c1">## A list of sample names (keys) and their corresponding FASTQ files (values)</span>
<span class="n">samples</span> <span class="o">=</span> <span class="n">OrderedDict</span><span class="p">({</span><span class="s1">&#39;ENCLB279NMT&#39;</span><span class="p">:</span> <span class="p">[</span><span class="s1">&#39;ENCFF824TZM&#39;</span><span class="p">,</span> <span class="s1">&#39;ENCFF176JNE&#39;</span><span class="p">]})</span>
<span class="n">fastq</span> <span class="o">=</span> <span class="n">paths</span><span class="p">([[</span><span class="n">f</span><span class="s2">&quot;{sample_dir}/{s}/{q}.fastq.gz&quot;</span> <span class="k">for</span> <span class="n">q</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">[</span><span class="n">s</span><span class="p">]]</span> <span class="k">for</span> <span class="n">s</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">])</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">default</span><span class="p">]</span>
<span class="n">sos_run</span><span class="p">(</span><span class="s1">&#39;align+call&#39;</span><span class="p">)</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Resource-preparation">Resource preparation<a class="anchor-link" href="#Resource-preparation">&#182;</a></h2><h3 id="hg19-human-reference-data">hg19 human reference data<a class="anchor-link" href="#hg19-human-reference-data">&#182;</a></h3><p>Obtain <code>hg19.fa</code> and <code>Homo_sapiens.GRCh38.91.gtf.gz</code>.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">hg19_reference_1</span> <span class="p">(</span><span class="n">download</span><span class="p">)]</span>
<span class="c1"># Download `hg19.2bit` and `twoBitToFa` from {ucsc_url}</span>
<span class="n">ucsc_url</span> <span class="o">=</span> <span class="s2">&quot;http://hgdownload.cse.ucsc.edu&quot;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/hg19.2bit&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/twoBitToFa&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">resource_dir</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="p">{</span><span class="n">ucsc_url</span><span class="p">}</span><span class="o">/</span><span class="n">goldenPath</span><span class="o">/</span><span class="n">hg19</span><span class="o">/</span><span class="n">bigZips</span><span class="o">/</span><span class="n">hg19</span><span class="o">.</span><span class="mi">2</span><span class="n">bit</span>
    <span class="p">{</span><span class="n">ucsc_url</span><span class="p">}</span><span class="o">/</span><span class="n">admin</span><span class="o">/</span><span class="n">exe</span><span class="o">/</span><span class="n">linux</span><span class="o">.</span><span class="n">x86_64</span><span class="o">/</span><span class="n">twoBitToFa</span>

<span class="p">[</span><span class="n">hg19_reference_2</span> <span class="p">(</span><span class="n">decompress</span> <span class="n">hg19</span><span class="o">.</span><span class="n">fa</span><span class="p">)]</span>
<span class="c1"># Use `twoBitToFa` to extract `hg19.fa` from `hg19.2bit`</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/{ref_fa}&quot;</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">chmod</span> <span class="o">+</span><span class="n">x</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span>
    <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">hg19_reference_3</span> <span class="p">(</span><span class="n">gene</span> <span class="n">annotations</span><span class="p">)]</span>
<span class="c1"># Download `Homo_sapiens.GRCh38.91.gtf.gz` from Ensembl</span>
<span class="c1"># https://useast.ensembl.org/info/data/ftp/index.html</span>
<span class="n">ensembl_ftp</span> <span class="o">=</span> <span class="s1">&#39;ftp://ftp.ensembl.org/pub/release-91/gtf/homo_sapiens/&#39;</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/{ref_gtf}&quot;</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">resource_dir</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="p">{</span><span class="n">ensembl_ftp</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_gtf</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Software-tools">Software tools<a class="anchor-link" href="#Software-tools">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">wasp</span><span class="p">]</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{wasp_dir}/README.md&quot;</span>
<span class="kn">download:</span> <span class="n">decompress</span> <span class="o">=</span> <span class="bp">True</span><span class="p">,</span> <span class="n">dest_file</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}/WASP.zip&#39;</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">github</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">bmvdgeijn</span><span class="o">/</span><span class="n">WASP</span><span class="o">/</span><span class="n">archive</span><span class="o">/</span><span class="n">master</span><span class="o">.</span><span class="n">zip</span>
<span class="kn">bash:</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">cwd</span><span class="p">}</span><span class="o">/</span><span class="n">WASP</span><span class="o">.</span><span class="n">zip</span>

<span class="p">[</span><span class="n">star</span><span class="p">]</span>
<span class="c1"># Quite time &amp; resource consuming (3hrs, 32GB memory)</span>
<span class="kn">depends:</span> <span class="n">Py_Module</span><span class="p">(</span><span class="s1">&#39;docker&#39;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s2">&quot;{resource_dir}/genomeParameters.txt&quot;</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;bschiffthaler/ngs&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
     <span class="n">STAR</span> <span class="o">--</span><span class="n">runMode</span> <span class="n">genomeGenerate</span> \
        <span class="o">--</span><span class="n">genomeDir</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">genomeFastaFiles</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_fa</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">sjdbGTFtagExonParentTranscript</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_gtf</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">runThreadN</span> <span class="p">{</span><span class="n">ncpu</span><span class="p">}</span>

<span class="p">[</span><span class="n">quasar</span><span class="p">]</span>
<span class="kn">depends:</span> <span class="n">R_library</span><span class="p">(</span><span class="s2">&quot;QuASAR@piquelab/QuASAR&quot;</span><span class="p">)</span>
<span class="kn">output:</span> <span class="n">f</span><span class="s1">&#39;{resource_dir}/convertPileupToQuasar.R&#39;</span>
<span class="kn">download:</span> <span class="n">dest_file</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{resource_dir}/convertPileupToQuasar.R&#39;</span>
    <span class="n">https</span><span class="p">:</span><span class="o">//</span><span class="n">raw</span><span class="o">.</span><span class="n">githubusercontent</span><span class="o">.</span><span class="n">com</span><span class="o">/</span><span class="n">piquelab</span><span class="o">/</span><span class="n">QuASAR</span><span class="o">/</span><span class="n">master</span><span class="o">/</span><span class="n">scripts</span><span class="o">/</span><span class="n">convertPileupToQuasar</span><span class="o">.</span><span class="n">R</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>Also apparently <code>WASP</code> only works with Python 2 ... need to create a <code>conda</code> environment for it:</p>

<pre><code>conda create -n py27 python=2.7
source activate py27
conda install pytables=2.4.0
pip install pysam</code></pre>
<p>and use</p>

<pre><code>source activate py27</code></pre>
<p>in <code>WASP</code> steps.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Get-samples">Get samples<a class="anchor-link" href="#Get-samples">&#182;</a></h2><p>FIXME: add description -- what are these samples?</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">obtain_samples</span><span class="p">]</span>
<span class="c1"># Download samples from ENCODE</span>
<span class="c1"># https://www.encodeproject.org/experiments/ENCSR384KAN/</span>
<span class="n">encode_url</span> <span class="o">=</span> <span class="s1">&#39;https://www.encodeproject.org/files&#39;</span>
<span class="kn">input:</span> <span class="kp">for_each</span> <span class="o">=</span> <span class="s1">&#39;fastq&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">fastq</span><span class="p">,</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">download:</span> <span class="n">dest_dir</span> <span class="o">=</span> <span class="n">sample_dir</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="p">{</span><span class="n">encode_url</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">_fastq</span><span class="p">:</span><span class="n">bnn</span><span class="p">}</span><span class="o">/</span><span class="err">@</span><span class="nd">@download</span><span class="o">/</span><span class="p">{</span><span class="n">_fastq</span><span class="p">:</span><span class="n">b</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Alignment">Alignment<a class="anchor-link" href="#Alignment">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Prefiltering-alignment">Prefiltering alignment<a class="anchor-link" href="#Prefiltering-alignment">&#182;</a></h3><p>Align with <code>STAR</code>, followd by <code>samtools</code> to remove reads with quality less than given cutoff.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">align_1</span> <span class="p">(</span><span class="n">STAR</span> <span class="n">prefiltering</span> <span class="n">alignment</span><span class="p">)]</span>
<span class="kn">parameter:</span> <span class="n">qual_cutoff</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">depends:</span> <span class="n">sos_step</span><span class="p">(</span><span class="s1">&#39;star&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="n">fastq</span><span class="p">,</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">2</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="p">[</span><span class="n">f</span><span class="s2">&quot;{sample_dir}/{x}.qual{qual_cutoff}.bam&quot;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">samples</span><span class="p">],</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;bschiffthaler/ngs&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesIn</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="o">--</span><span class="n">readFilesCommand</span> <span class="n">zcat</span> \
        <span class="o">--</span><span class="n">runThreadN</span> <span class="p">{</span><span class="n">ncpu</span><span class="p">}</span> <span class="o">--</span><span class="n">outStd</span> <span class="n">BAM_SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">sjdbGTFtagExonParentTranscript</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_gtf</span><span class="p">}</span> <span class="o">|</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">bq</span> <span class="p">{</span><span class="n">qual_cutoff</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="WASP-informed-remap">WASP-informed remap<a class="anchor-link" href="#WASP-informed-remap">&#182;</a></h3><p>Remap reads after <a href="https://github.com/bmvdgeijn/WASP/tree/master/mapping"><code>WASP</code></a> adjustment. <code>WASP</code> uses pre-defined list of SNPs and removes bias caused by them. The list was generated by Yanyu using 1k genome SNP in VCF format with MAF filter as input to <code>WASP/mapping/extract_vcf_snps.sh</code> command. Here I just take this pre-compiled list from Yanyu.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">align_2</span> <span class="p">(</span><span class="n">WASP</span> <span class="n">intersecting</span> <span class="n">SNP</span><span class="p">):</span> <span class="kp">shared</span> <span class="o">=</span> <span class="p">{</span><span class="s1">&#39;wasp_split&#39;</span><span class="p">:</span> <span class="s1">&#39;step_output&#39;</span><span class="p">}]</span>
<span class="c1"># WASP finding unbiased reads intersecting with SNP</span>
<span class="kn">depends:</span> <span class="n">sos_step</span><span class="p">(</span><span class="s1">&#39;wasp&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="p">[[</span><span class="n">f</span><span class="s2">&quot;{x:n}.remap.fq1.gz&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{x:n}.remap.fq2.gz&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{x:n}.to.remap.bam&quot;</span><span class="p">,</span> <span class="n">f</span><span class="s2">&quot;{x:n}.keep.bam&quot;</span><span class="p">]</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="n">_input</span><span class="p">],</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">4</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">source</span> <span class="n">activate</span> <span class="n">py27</span>
    <span class="n">python</span> <span class="p">{</span><span class="n">wasp_dir</span><span class="p">}</span><span class="o">/</span><span class="n">find_intersecting_snps</span><span class="o">.</span><span class="n">py</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">snp_dir</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">wasp_snp_list</span> \
        <span class="o">--</span><span class="n">is_sorted</span> <span class="o">--</span><span class="n">is_paired_end</span>

<span class="p">[</span><span class="n">align_3</span> <span class="p">(</span><span class="n">STAR</span> <span class="n">post</span> <span class="n">alignment</span><span class="p">)]</span>
<span class="c1"># Align WASP remap with STAR</span>
<span class="c1"># Followd by samtools remove reads with quality less than {qual_cutoff}</span>
<span class="kn">parameter:</span> <span class="n">qual_cutoff</span> <span class="o">=</span> <span class="mi">10</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">4</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.{qual}.remap.{ext}&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.remapped.qual{qual_cutoff}.bam&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;bschiffthaler/ngs&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
    <span class="n">STAR</span> <span class="o">--</span><span class="n">genomeDir</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span> \
        <span class="o">--</span><span class="n">readFilesIn</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">0</span><span class="p">]}</span> <span class="p">{</span><span class="n">_input</span><span class="p">[</span><span class="mi">1</span><span class="p">]}</span> \
        <span class="o">--</span><span class="n">readFilesCommand</span> <span class="n">zcat</span> \
        <span class="o">--</span><span class="n">runThreadN</span> <span class="p">{</span><span class="n">ncpu</span><span class="p">}</span> <span class="o">--</span><span class="n">outStd</span> <span class="n">BAM_SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">outSAMtype</span> <span class="n">BAM</span> <span class="n">SortedByCoordinate</span> \
        <span class="o">--</span><span class="n">sjdbGTFtagExonParentTranscript</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_gtf</span><span class="p">}</span> <span class="o">|</span>
    <span class="n">samtools</span> <span class="n">view</span> <span class="o">-</span><span class="n">bq</span> <span class="p">{</span><span class="n">qual_cutoff</span><span class="p">}</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">align_4</span> <span class="p">(</span><span class="n">WASP</span> <span class="n">remove</span> <span class="n">ambiguously</span> <span class="n">mapped</span> <span class="n">reads</span><span class="p">)]</span>
<span class="n">to_remap</span> <span class="o">=</span> <span class="n">paths</span><span class="p">([</span><span class="n">wasp_split</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">4</span><span class="p">][</span><span class="mi">2</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">wasp_split</span><span class="p">),</span> <span class="mi">4</span><span class="p">)])</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">paired_with</span> <span class="o">=</span> <span class="s1">&#39;to_remap&#39;</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.remapped.{ext}&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.keep_remapped.{_ext[0]}&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.log&#39;</span>
    <span class="n">source</span> <span class="n">activate</span> <span class="n">py27</span>
    <span class="n">python</span> <span class="p">{</span><span class="n">wasp_dir</span><span class="p">}</span><span class="o">/</span><span class="n">filter_remapped_reads</span><span class="o">.</span><span class="n">py</span> <span class="p">{</span><span class="n">_to_remap</span><span class="p">}</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">align_5</span> <span class="p">(</span><span class="n">Merge</span> <span class="n">WASP</span> <span class="n">adjusted</span> <span class="ow">and</span> <span class="n">originally</span> <span class="n">kept</span> <span class="n">BAM</span><span class="p">)]</span>
<span class="n">kept</span> <span class="o">=</span> <span class="n">paths</span><span class="p">([</span><span class="n">wasp_split</span><span class="p">[</span><span class="n">i</span><span class="p">:</span><span class="n">i</span><span class="o">+</span><span class="mi">4</span><span class="p">][</span><span class="mi">3</span><span class="p">]</span> <span class="k">for</span> <span class="n">i</span> <span class="ow">in</span> <span class="nb">range</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span> <span class="nb">len</span><span class="p">(</span><span class="n">wasp_split</span><span class="p">),</span> <span class="mi">4</span><span class="p">)])</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">paired_with</span> <span class="o">=</span> <span class="s1">&#39;kept&#39;</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.keep_remapped.{ext}&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.wasp_remapped.{_ext[0]}&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;bschiffthaler/ngs&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
    <span class="n">samtools</span> <span class="n">merge</span> <span class="o">-</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="p">{</span><span class="n">_kept</span><span class="p">}</span> <span class="o">|</span> <span class="n">samtools</span> <span class="n">sort</span> <span class="o">-</span><span class="n">o</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
    <span class="n">samtools</span> <span class="n">index</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">align_6</span> <span class="p">(</span><span class="n">WASP</span> <span class="n">remove</span> <span class="n">duplicate</span> <span class="n">reads</span> <span class="p">)]</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.bam&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.dedup.bam&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{_output:n}.log&#39;</span>
    <span class="n">source</span> <span class="n">activate</span> <span class="n">py27</span>
    <span class="n">python</span> <span class="p">{</span><span class="n">wasp_dir</span><span class="p">}</span><span class="o">/</span><span class="n">rmdup</span><span class="o">.</span><span class="n">py</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

</div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Genotype-and-ASE-calling">Genotype and ASE calling<a class="anchor-link" href="#Genotype-and-ASE-calling">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div>
<div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Pre-QuASAR-calling">Pre-<code>QuASAR</code> calling<a class="anchor-link" href="#Pre-QuASAR-calling">&#182;</a></h3><p>This step extracts count information from BAM file based on SNP list and prepare it into <code>QuASAR</code> input format. Here we need the list of SNPs used in BED format, and reference genome <code>fasta</code> file. Yanyu has previously prepared this BED file. Here I just grab and use it.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">call_1</span> <span class="p">(</span><span class="n">samtools</span> <span class="n">pileup</span><span class="p">)]</span>
<span class="c1"># A standard samtools mpileup call</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.bam&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.pileup.gz&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;bschiffthaler/ngs&#39;</span><span class="p">,</span> <span class="n">expand</span><span class="o">=</span><span class="bp">True</span>
    <span class="n">samtools</span> <span class="n">mpileup</span> <span class="o">-</span><span class="n">f</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="p">{</span><span class="n">ref_fa</span><span class="p">}</span> <span class="o">-</span><span class="n">l</span> <span class="p">{</span><span class="n">samtools_snp_list</span><span class="p">}</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">|</span> <span class="n">bgzip</span> <span class="o">&gt;</span> <span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">call_2</span> <span class="p">(</span><span class="n">awk</span> <span class="o">&amp;</span> <span class="n">bedtools</span> <span class="nb">filter</span> <span class="n">pileup</span><span class="p">)]</span>
<span class="c1"># Remove bad pileup and intersect only with list of SNPs provided</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.pileup.gz&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.pileup.bed.gz&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">docker_image</span> <span class="o">=</span> <span class="s1">&#39;biocontainers/bedtools&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span>
    <span class="n">cat</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">|</span> \
    <span class="n">awk</span> <span class="o">-</span><span class="n">v</span> <span class="n">OFS</span><span class="o">=</span><span class="s1">&#39;</span><span class="se">\\</span><span class="s1">t&#39;</span> <span class="s1">&#39;{ if ($4&gt;0 &amp;&amp; $5 !~ /</span><span class="se">\\</span><span class="s1">+[0-9]+[ACGTNacgtn]+/ &amp;&amp; $5 !~ /-[0-9]+[ACGTNacgtn]+/ &amp;&amp; $5 !~ /[^</span><span class="se">\\</span><span class="s1">^]</span><span class="se">\\</span><span class="s1">*/) print $1,$2-1,$2,$3,$4,$5,$6}&#39;</span> <span class="o">|</span> \
    <span class="n">sortBed</span> <span class="o">-</span><span class="n">i</span> <span class="n">stdin</span> <span class="o">|</span> <span class="n">intersectBed</span> <span class="o">-</span><span class="n">a</span> <span class="n">stdin</span> <span class="o">-</span><span class="n">b</span> <span class="err">$</span><span class="p">{</span><span class="n">samtools_snp_list</span><span class="p">}</span> <span class="o">-</span><span class="n">wo</span> <span class="o">|</span> \
    <span class="n">cut</span> <span class="o">-</span><span class="n">f</span> <span class="mi">1</span><span class="o">-</span><span class="mi">7</span><span class="p">,</span><span class="mi">11</span><span class="o">-</span><span class="mi">14</span> <span class="o">|</span> <span class="n">gzip</span> <span class="o">&gt;</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>

<span class="p">[</span><span class="n">call_3</span> <span class="p">(</span><span class="n">convert</span> <span class="n">to</span> <span class="n">quasar</span> <span class="n">format</span><span class="p">)]</span>
<span class="kn">depends:</span> <span class="n">sos_step</span><span class="p">(</span><span class="s1">&#39;quasar&#39;</span><span class="p">)</span>
<span class="kn">input:</span> <span class="kp">group_by</span> <span class="o">=</span> <span class="mi">1</span><span class="p">,</span> <span class="kp">pattern</span> <span class="o">=</span> <span class="s1">&#39;{name}.pileup.bed.gz&#39;</span><span class="p">,</span> <span class="kp">concurrent</span> <span class="o">=</span> <span class="bp">True</span>
<span class="kn">output:</span> <span class="n">expand_pattern</span><span class="p">(</span><span class="n">f</span><span class="s1">&#39;{_name[0]}.quasar.in.gz&#39;</span><span class="p">)</span>
<span class="kn">bash:</span> <span class="kp">workdir</span> <span class="o">=</span> <span class="n">f</span><span class="s1">&#39;{cwd:a}&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="bp">True</span>
    <span class="n">R</span> <span class="o">--</span><span class="n">slave</span> <span class="o">--</span><span class="n">args</span> <span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">&lt;</span> <span class="p">{</span><span class="n">resource_dir</span><span class="p">}</span><span class="o">/</span><span class="n">convertPileupToQuasar</span><span class="o">.</span><span class="n">R</span>
</pre></div>

</div>
</div>
</div>

</div>
<hr>
&copy 2018 Min Qiao at He Lab, University of Chicago
<p><small>Exported from <a href="http://github.com/gaow/m6a/blob/36bb1fe236a40c88a678cce9a1106846c1161cd8/analysis/20180311_ASE_Calling_WASP_QuASAR.ipynb"><code>analysis/20180311_ASE_Calling_WASP_QuASAR.ipynb</code></a> committed by Min Qiao on Tue Mar 13 23:21:46 2018 <a href="http://github.com/gaow/m6a/commit/36bb1fe236a40c88a678cce9a1106846c1161cd8">revision 10, 36bb1fe</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
